File: contentGeneration/ingest.py
import os
import shutil
import json
from langchain_community.document_loaders import JSONLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_huggingface import HuggingFaceEmbeddings
from langchain_community.vectorstores import Chroma

# Define the path for the ChromaDB database and the source documents
CHROMA_PATH = "chroma_db"
DATA_PATH = "data/"

def metadata_func(record: dict, metadata: dict) -> dict:
    """Helper function to extract metadata from a JSON record."""
    metadata["source"] = record.get("source_url")
    return metadata

def create_vector_db():
    """
    Creates a ChromaDB vector database from JSON documents in the data directory.
    This version intelligently handles different JSON structures.
    """
    print("--- Starting JSON Document Ingestion Process with ChromaDB ---")
    
    all_documents = []

    print("Loading JSON files...")
    # Loop through all files in the data directory
    for filename in os.listdir(DATA_PATH):
        if filename.endswith(".json"):
            filepath = os.path.join(DATA_PATH, filename)
            
            # --- THIS IS THE KEY FIX ---
            # Determine the correct jq_schema by checking the JSON structure
            with open(filepath, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            # If the file contains a list of objects, iterate over the list.
            # Otherwise, treat it as a single object.
            if isinstance(data, list):
                jq_schema = ".[]"
                print(f"Detected list structure in {filename}, using schema '.[]'")
            else:
                jq_schema = "."
                print(f"Detected single object structure in {filename}, using schema '.'")
            # --- END OF FIX ---
            
            # Now, create the loader with the correct schema
            loader = JSONLoader(
                file_path=filepath,
                jq_schema=jq_schema,
                content_key="content",
                metadata_func=metadata_func,
                text_content=False
            )
            all_documents.extend(loader.load())

    if not all_documents:
        print("No documents found. Please add JSON content to the 'data' directory.")
        return
            
    print(f"Loaded a total of {len(all_documents)} document(s)/records.")

    # 2. Split Text into Chunks (No changes here)
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=1000,
        chunk_overlap=100
    )
    texts = text_splitter.split_documents(all_documents)
    print(f"Split documents into {len(texts)} chunks.")

    # 3. Load the Embedding Model (No changes here)
    embeddings = HuggingFaceEmbeddings(
        model_name='sentence-transformers/all-MiniLM-L6-v2',
        model_kwargs={'device': 'cpu'}
    )
    print("Embedding model loaded.")

    # 4. Create and Persist the ChromaDB Database (No changes here)
    if os.path.exists(CHROMA_PATH):
        print(f"Clearing existing database at {CHROMA_PATH}")
        shutil.rmtree(CHROMA_PATH)

    db = Chroma.from_documents(
        documents=texts, 
        embedding=embeddings, 
        persist_directory=CHROMA_PATH
    )
    
    print(f"--- Vector store created and saved at: {CHROMA_PATH} ---")
    print(f"Number of vectors in the database: {db._collection.count()}")

if __name__ == "__main__":
    create_vector_db()

File: ingest.ts
import { JSONLoader } from 'langchain/document_loaders/fs/json';
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
import { Chroma } from '@langchain/community/vectorstores/chroma';
import { GoogleGenerativeAIEmbeddings } from '@langchain/google-genai';
import dotenv from 'dotenv'

import * as fs from 'fs';
import * as path from 'path';
dotenv.config()
// --- CONFIGURATION ---
// The path to your source content files
const SOURCE_DATA_PATH = './contentGeneration/chroma_db'; 
// The URL of your running ChromaDB server (from docker-compose)
const CHROMA_URL = 'http://localhost:8000';
// The name of the collection within Chroma where the data will be stored
const CHROMA_COLLECTION_NAME = 'ecolearn_content';

/**
 * The main function to run the ingestion process.
 */
const run = async () => {
  try {
    console.log('--- Starting Ingestion Process ---');

    // 1. Load Documents from JSON files using a robust method
    const docs = [];
    const files = fs.readdirSync(SOURCE_DATA_PATH).filter(file => file.endsWith('.json'));

    if (files.length === 0) {
        console.log('No JSON documents found in the data directory. Exiting.');
        return;
    }

    console.log(process.env.GOOGLE_API_KEY,"sghfjksdbfjksdb");
    console.log(`Found ${files.length} JSON file(s). Loading content...`);

    for (const file of files) {
        const filePath = path.join(SOURCE_DATA_PATH, file);
        // The JSONLoader can handle both single objects and lists of objects
        // The '/content' is a JSON Path expression to extract the value of the "content" key.
        const loader = new JSONLoader(filePath, '/content'); 
        
        const loadedDocs = await loader.load();
        docs.push(...loadedDocs);
    }

    console.log(`Loaded a total of ${docs.length} document(s)/records.`);

    // 2. Split Text into Chunks
    const textSplitter = new RecursiveCharacterTextSplitter({
      chunkSize: 1000,
      chunkOverlap: 100,
    });
    const splitDocs = await textSplitter.splitDocuments(docs);
    console.log(`Split documents into ${splitDocs.length} chunks.`);
    // 3. Initialize the Embedding Model
    const embeddings = new GoogleGenerativeAIEmbeddings({
    apiKey: process.env.GOOGLE_API_KEY,
      model: 'embedding-001',
    });
    console.log('Google Generative AI Embeddings model loaded.');

    // 4. Create and Persist the ChromaDB Vector Store
    console.log(`Connecting to ChromaDB at ${CHROMA_URL} and adding documents to collection '${CHROMA_COLLECTION_NAME}'...`);
    
    // This will connect to your running Chroma instance and add the documents.
    // The persistence is handled by the Docker volume.
    await Chroma.fromDocuments(splitDocs, embeddings, {
      collectionName: CHROMA_COLLECTION_NAME,
      url: CHROMA_URL,
    });

    console.log('--- Ingestion Process Finished Successfully ---');

  } catch (error) {
    console.error('An error occurred during ingestion:', error);
    process.exit(1); // Exit with an error code
  }
};

run();